tensorflow/compiler/xla/service/algebraic_simplifier.cc

```diff
@@ -1665,6 +1668,228 @@ StatusOr<HloInstruction*> AlgebraicSimplifierVisitor::OptimizeDotOfGather(
   return memoized_lookup;
 }
 
+// This function tries to transform
+//   dot(reshape(transpose(A)), Const) to
+//   dot(reshape(A), reshape(transpose(reshape(Const)))),
+// so that the reshape and transpose on the Const side can be constant folded.
+//
+// The basic idea is that since the accumulation in the dot operation is
+// associative, so as long as we permute the elements of the contracting
+// dimensions on both sides of the dot in the same way, the result of the
+// dot is not affected.
+StatusOr<HloInstruction*>
+AlgebraicSimplifierVisitor::OptimizeDotOfReorderContractingDims(
+    HloInstruction* dot) {
+  // This transformation assumes layout is not assigned yet.
+  if (options_.is_layout_sensitive()) {
+    return nullptr;
+  }
+
+  // Canonicalize dot(<constant>, rhs) to dot(rhs, <constant>) to make the
+  // remainder of this function easier.
+  auto dnums = dot->dot_dimension_numbers();
+  auto lhs_contracting_dims = dnums.lhs_contracting_dimensions();
+  auto rhs_contracting_dims = dnums.rhs_contracting_dimensions();
+  auto* lhs = dot->mutable_operand(0);
+  auto* rhs = dot->mutable_operand(1);
+  if (dot->operand(0)->IsConstant()) {
+    std::swap(lhs, rhs);
+    std::swap(lhs_contracting_dims, rhs_contracting_dims);
+  }
+
+  // Require single contracting dim to make the implementation easier to
+  // track contracting dims.
+  if (dnums.lhs_contracting_dimensions_size() != 1) {
+    return nullptr;
+  }
+
+  // Pattern match Dot(reshape(transpose(input), constant))
+  HloInstruction* reshape;
+  HloInstruction* transpose;
+  HloInstruction* input;
+  HloInstruction* constant;
+  if (!Match(lhs,
+             m::Reshape(&reshape, m::Transpose(&transpose, m::Op(&input)))) ||
+      !Match(rhs, m::Constant(&constant))) {
+    return nullptr;
+  }
+
+  // Check if reshape squishes some dims into one dim, and that if this one
+  // dim is the dot's lhs contracting dim.
+  CHECK(lhs_contracting_dims.size() == 1);
+  const std::vector<std::pair<int64, int64>>& unmodified_dims =
+      ShapeUtil::DimensionsUnmodifiedByReshape(lhs->operand(0)->shape(),
+                                               lhs->shape());
+  if (unmodified_dims.size() != lhs->shape().rank() - 1) {
+    return nullptr;
+  }
+  for (const auto &pair : unmodified_dims) {
+    if (pair.second == lhs_contracting_dims[0]) {
+      // If the contracting dim is unmodified, and only one dim is modified,
+      // then that dim must be a non-contracting dim.
+      return nullptr;
+    }
+  }
```

Comment:
This may be clearer as a `absl::c_linear_search` over unmodified_dims?

Reply:
Agreed, will do.

Reply:
Since we only check the pair.second == something, we need a custom compare function. So I change it to std::find_if.